### DISCUSSION TITLE
Graceful shutdown

### DISCUSSION THREAD

**Discussion by PepijnSenders** [vercel/next.js#19693] [Ideas]
# Feature request

When next.js shuts down, it takes care of exiting the current process with code `0`. This function does not close the HTTP server before it [exits the process](https://github.com/vercel/next.js/blob/397a375b37ca7727f13d73b36fcc80df61cfc66a/packages/next/bin/next.ts#L108), next.js should make sure [the server has been `.close()`d before exiting](https://nodejs.org/api/http.html#http_server_close_callback). We also need something for the case of people opening database connections and these being closed before the process exits.

This behavior is problematic in environments like kubernetes because the server is immediately shut down and will still be receiving requests as kubernetes replaces the pod. How this works is described quite well in detail [here](https://blog.dashlane.com/implementing-nodejs-http-graceful-shutdown/). We need to make sure that we can expose an endpoint where we signal that the pod should not be receiving any more requests, and make sure that after a set timeout the server gracefully shuts down. This is quite easy to implement with libraries like [terminus](https://github.com/godaddy/terminus).

## Is your feature request related to a problem? Please describe.

Next.js doesn't shut down properly.

## Describe the solution you'd like

A graceful shutdown solution to be added to the core project, with hooks to be able to shut down things like database connections.

## Describe alternatives you've considered

This is easily fixed by adding a custom server, but I believe that this should be part of the core next.js code.

**Comment by banksalad-seongwoo** [comment]
To resolve it, I just opend this https://github.com/vercel/next.js/pull/29959

**Reply by airtoxin** [reply]
The **graceful** means, Stop accepting newly http connection, and completely end http request then shutting down server.
It is a good idea to execute the `server.close()` function after receiving the SIGTERM signal.
https://nodejs.org/api/net.html#serverclosecallback

**Reply by EmilioAiolfi** [reply]
@ijjk a doubt, is it possible to use "server.close()" of "Next Server"?
A new version 12.2.0 helps, but how I can be use server.close() ou another approach

```js
// pages/_document.js
if (process.env.NEXT_MANUAL_SIG_HANDLE) {
  // this should be added in your custom _document
  process.on('SIGTERM', () => {
    console.log('Received SIGTERM: ', 'cleaning up')
    process.exit(0)
  })
  process.on('SIGINT', () => {
    console.log('Received SIGINT: ', 'cleaning up')
    process.exit(0)
  })
}
```
```

**Reply by finkrer** [reply]
@ijjk After #36909 we can write our own hooks for SIG events, but there's still no way to gracefully shutdown the Next server itself. Next just doesn't expose the server unless you use a custom one.

I really think the Next server should do this on its own, by default — without any user hooks. We just found that Next is leaving requests unanswered during restarts, most people probably won't notice that, but the problem is still there.

**Reply by vamcs** [reply]
@finkrer it sounds like you need to configure the keep-alive timeout for the server, take a look: https://nextjs.org/docs/api-reference/cli#keep-alive-timeout. 

I'm also investigating how to correctly implement graceful shutdowns without a custom server, and I happened to find that page in the docs. We use a load balancer so I believe it should be enough for my case since we don't open DB connections or do anything that needs cleaning up before closing the server.

**Reply by finkrer** [reply]
Hi @vamcs. 

As far as I understand, this setting would not help. The issue is that upon receiving the terminate signal Next just quits, while the timeout setting affects Next timing out connections while it's running. 

Besides, I have tried using this setting before just in case since we are also using a reverse proxy, and it just broke the application outright.  I think we were getting 502 errors from nginx.

**Comment by jmkng** [comment]
Using Next.js for a full stack project is hard to stomach when simple stuff like this is hard to access. This is just a big red flag.

**Comment by azerum** [comment]
I though that calling `process.exit()` if often unnecessary, as Node should shutdown by itself when there's no pending work left. It's also mentioned in the [Node docs](https://nodejs.org/api/process.html#processexitcode):

> Calling process.exit() will force the process to exit as quickly as possible even if there are still asynchronous operations pending that have not yet completed fully, including I/O operations to process.stdout and process.stderr.
>
> In most situations, it is not actually necessary to call process.exit() explicitly. The Node.js process will exit on its own if there is no additional work pending in the event loop. The process.exitCode property can be set to tell the process which exit code to use when the process exits gracefully.

Is there any specific reason why Next uses `process.exit()` instead of graceful shutdown?

**Comment by CaseyHaralson** [comment]
It looks like this commit might have implemented it: https://github.com/vercel/next.js/commit/c5a8e0989ea13af8d04bf6dda8d741c79b964abb (packages/next/src/server/lib/start-server.ts line 192)

**Reply by nebarf** [reply]
Looking at the implementation it does call `process.exit` immediately after `server.close` without waiting for existing connections to be ended. `server.close` is async (doc https://nodejs.org/api/net.html#serverclosecallback)

**Comment by lukevers** [comment]
Linking my comment here from another thread I added -- https://github.com/vercel/next.js/issues/51404#issuecomment-1912701275

This works locally, but I haven't tried it in production yet:

```js
const process = require('process');
const spawn = require('child_process').spawn;

// Adding `detatched: true` makes sure the signal doesn't send down automatically
const server = spawn('pnpm', ['run', 'start'], { detached: true });
server.stdout.on('data', function(data) {
  process.stdout.write(data);
});

server.stderr.on('data', function (data) {
  process.stderr.write(data);
});

server.on('exit', function (code) {
  console.log('Application exited with code:', code.toString());
});

['SIGTERM', 'SIGINT'].forEach((signal) => {
  process.on(signal, function (signal) {
    console.log('Caught signal', signal, '-- Allowing requests to drain');

    setTimeout(() => {
      console.log('Terminating Application')
      server.kill(signal);
    }, 15 * 1000);

    setTimeout(function() {
      console.log('Terminating Wrapper');
      process.exit(0);
    }, 20 * 1000);
  });
});
```

**Comment by ryuichi1208** [comment]
Do you think the issue discussed in this Discussion has been resolved by the merge of the following PR? https://github.com/vercel/next.js/pull/60059

**Reply by novascreen** [reply]
Ideally we'd have a configurable delay. In Kubernetes we need to wait until the readiness probe fails and we no longer get new traffic, then we can let `server.close` do its thing.

**Reply by novascreen** [reply]
Rather than a timeout i think a good solution might be to allow us to do something like this
```ts
  process.on('SIGINT', () => {
	// timeout, cleanup...
    delete process.env.NEXT_MANUAL_SIG_HANDLE
    process.emit('SIGINT')
  })
```
But for that to work the current `NEXT_MANUAL_SIG_HANDLE` check needs to be inside the cleanup handler rather than a check before setting up the event handlers.

Otherwise we have no way to trigger the actual graceful shutdown.

https://github.com/vercel/next.js/pull/82449#discussion_r2277027099

**Comment by Ashvith10** [comment]
I need this to run cleanup functions used by mikroORM. Any solution?
